helix-core/src/main/java/org/apache/helix/controller/rebalancer/strategy/CrushRebalanceStrategy.java

```diff
@@ -89,7 +89,16 @@ public ZNRecord computePartitionAssignment(final List<String> allNodes,
       long data = partitionName.hashCode();
 
       // apply the placement rules
-      List<Node> selected = select(topNode, data, _replicas, eventId);
+      List<Node> selected;
+      try {
+        selected = select(topNode, data, _replicas, eventId);
+      } catch (IllegalStateException e) {
+        String errorMessage = String
+            .format("Could not select enough number of nodes. %s partition %s, required %d",
+                _resourceName, partitionName, _replicas);
+        LogUtil.logError(Log, eventId, errorMessage);
```

Comment:
Since you are throwing the exception, would it be better not to log the error message twice? If you log the exception and then throw it, there’s a chance that the exception will be logged again further up the call stack, resulting in two log events for the same error. I would add eventId into the errerMessage would be a good idea.

Reply:
I think it better to read the code completely or you have some understanding of Helix pipeline. This exception will be handled in BestPossibleStage. It was designed for not breaking the pipeline which will impact the computation for rest of other resources.